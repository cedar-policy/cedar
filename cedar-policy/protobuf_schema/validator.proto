//
// Copyright Cedar Contributors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

syntax = "proto3";
package cedar_policy_validator;

import "core.proto";

// the protobuf Schema message describes a complete schema.
message Schema {
    // TODO: this need not be a map at all, since `EntityDecl` contains the `name` as well.
    // It can be just `repeated EntityDecl`.
    repeated EntityTypeToEntityDeclMap entity_decls = 1;
    // TODO: this need not be a map at all, since `ActionDecl` contains the `name` as well.
    // It can be just `repeated ActionDecl`.
    repeated EntityUidToActionDeclMap action_decls = 2;
}

// This `message` with `key` and `value`, rather than a `map`, since messages can't be dictionary keys
message EntityTypeToEntityDeclMap {
    cedar_policy_core.Name key = 1;
    EntityDecl value = 2;
}

// This `message` with `key` and `value`, rather than a `map`, since messages can't be dictionary keys
message EntityUidToActionDeclMap {
    cedar_policy_core.EntityUid key = 1;
    ActionDecl value = 2;
}

// the protobuf EntityDecl message contains all of the schema's
// information about a single entity type.
message EntityDecl {
    cedar_policy_core.Name name = 1;
    repeated cedar_policy_core.Name descendants = 2;
    map<string, AttributeType> attributes = 3;
    OpenTag open_attributes = 4;
    optional Type tags = 5;
    repeated string enum_choices = 6;
}

// the protobuf ActionDecl message contains all of the schema's
// information about a single action.
message ActionDecl {
    cedar_policy_core.EntityUid name = 1;
    repeated cedar_policy_core.EntityUid descendants = 3;
    Type context = 4;
    repeated cedar_policy_core.Name principal_types = 7;
    repeated cedar_policy_core.Name resource_types = 8;
}

message Type {
    oneof data {
        Ty ty = 1;
        Type set_type = 2;
        EntityRecordKind entityOrRecord = 3;
        cedar_policy_core.Name name = 4;
    }

    enum Ty {
        Never = 0;
        True = 1;
        False = 2;
        EmptySetType = 3;
        Bool = 4;
        String = 5;
        Long = 6;
    }
}

message EntityRecordKind {
    oneof data {
        AnyEntity any_entity = 1;
        Record record = 2;
        cedar_policy_core.Name entity = 3;
        ActionEntity actionEntity = 4;
    }

    // Zero-arity constructors represented as enums with only one member
    enum AnyEntity {
        // the one option for the enum
        unit = 0;
    }
    message Record {
        map<string, AttributeType> attrs = 1;
        OpenTag open_attributes = 2;
    }
    message ActionEntity {
        cedar_policy_core.Name name = 1;
        map<string, AttributeType> attrs = 2;
    }
}

enum OpenTag {
    OpenAttributes = 0;
    ClosedAttributes = 1;
}

message AttributeType {
    Type attr_type = 1;
    bool is_required = 2;
}

enum ValidationMode {
    Strict = 0;
    Permissive = 1;
}
