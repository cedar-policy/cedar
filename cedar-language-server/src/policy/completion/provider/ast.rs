/*
 * Copyright Cedar Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::sync::Arc;

use cedar_policy_core::{
    ast::{BinaryOp, EntityType, Expr, ExprVisitor, Literal, Pattern},
    parser::Loc,
};
use smol_str::SmolStr;

use crate::{
    policy::{
        context::{
            AttrContext, AttrContextKind, BinaryOpContext, InKind, IsContext, Op, ReceiverContext,
        },
        CompletionContextKind, DocumentContext,
    },
    utils::is_cursor_in_condition_braces,
};

/// A visitor implementation for gathering completion information from Cedar policy expressions.
///
/// This visitor traverses the Cedar AST to identify the expression context at the cursor position
/// within policy condition blocks(the `when/unless {...}` section of a policy) and determines what kind
/// of completions should be offered to the user.
pub(crate) struct ConditionCompletionVisitor<'a> {
    cx: &'a DocumentContext<'a>,
}

impl<'a> ConditionCompletionVisitor<'a> {
    /// Creates a new completion visitor with the given document context.
    fn new(document_context: &'a DocumentContext<'_>) -> Self {
        Self {
            cx: document_context,
        }
    }

    /// Analyzes the current policy and cursor position to determine the appropriate completion context.
    ///
    /// This is the main entry point for obtaining completion suggestions based on where the
    /// cursor is positioned within a Cedar policy's conditions.
    pub(crate) fn get_completion_context(
        document_context: &DocumentContext<'_>,
    ) -> CompletionContextKind {
        let expr = document_context.policy.non_scope_constraints();
        if !is_cursor_in_condition_braces(
            document_context.cursor_position,
            document_context.policy_text,
        ) {
            return CompletionContextKind::None;
        }

        let mut visitor = ConditionCompletionVisitor::new(document_context);
        let kind = visitor
            .visit_expr(expr)
            .unwrap_or(CompletionContextKind::Unknown);

        let word = document_context
            .get_word_under_cursor()
            // Filters out the typing of entity euid literals as to not provide suggestions for it
            .filter(|w| !w.contains(':'));

        match kind {
            // Must type a character to get suggestions within empty when blocks or after a completion
            CompletionContextKind::Unknown if word.is_some() => {
                CompletionContextKind::ConditionDefault
            }
            kind => kind,
        }
    }

    fn visit_attr(
        &mut self,
        expr: &Arc<Expr>,
        attr: &SmolStr,
        loc: Option<&Loc>,
        context: AttrContextKind,
    ) -> Option<<Self as ExprVisitor>::Output> {
        self.visit_expr(expr).or_else(|| {
            let word = self.cx.get_word_under_cursor();
            // If cursor is within attribute and matches the current attribute name (or empty attribute)
            if self.cx.is_cursor_over_loc(loc) && (attr.is_empty() || word == Some(attr)) {
                return CompletionContextKind::Attr(AttrContext::new(
                    ReceiverContext::new(expr.clone()),
                    context,
                ))
                .into();
            }
            None
        })
    }
}

impl ExprVisitor for ConditionCompletionVisitor<'_> {
    type Output = CompletionContextKind;

    fn visit_literal(
        &mut self,
        lit: &cedar_policy_core::ast::Literal,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        if let Some(loc) = loc {
            // Check if cursor is within the literal
            if self.cx.is_cursor_over_loc(loc) {
                // We're specifically interested in EntityUID literals for action completions
                if let Literal::EntityUID(entity_uid) = lit {
                    let word = self.cx.get_word_under_cursor();
                    if word == Some(entity_uid.eid().escaped().as_str()) {
                        return CompletionContextKind::EntityLiteral {
                            entity_uid: entity_uid.clone(),
                        }
                        .into();
                    }
                }

                if let Literal::String(..) = lit {
                    return CompletionContextKind::None.into();
                }
            }
        } else {
            // Fall back to entity ID location for action completion in policy scope
            if let Literal::EntityUID(entity_uid) = lit {
                return self
                    .visit_literal(&Literal::EntityUID(entity_uid.clone()), entity_uid.loc());
            }
        }
        None
    }

    fn visit_is(
        &mut self,
        expr: &Arc<Expr>,
        entity_type: &EntityType,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        self.visit_expr(expr).or_else(|| {
            if let Some(loc) = loc {
                if self.cx.is_cursor_over_loc(loc) {
                    let word = self.cx.get_word_under_cursor();
                    if word == Some(&entity_type.to_string()) {
                        return CompletionContextKind::Is(IsContext::new(ReceiverContext::new(
                            expr.clone(),
                        )))
                        .into();
                    }
                }
            }
            None
        })
    }

    fn visit_binary_op(
        &mut self,
        op: cedar_policy_core::ast::BinaryOp,
        arg1: &Arc<Expr>,
        arg2: &Arc<Expr>,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        // Try to get completions from either operand first
        self.visit_expr(arg1)
            .or_else(|| self.visit_expr(arg2))
            .or_else(|| {
                if let Some(loc) = loc {
                    // Determine which side needs completion
                    let complete_right_side = !arg1
                        .source_loc()
                        .is_some_and(|loc| self.cx.is_cursor_over_loc(loc));

                    let (other_side_expr, complete_side_expr) = if complete_right_side {
                        (arg1.clone(), arg2.clone())
                    } else {
                        (arg2.clone(), arg1.clone())
                    };

                    if self.cx.is_cursor_over_loc(loc) {
                        match op {
                            BinaryOp::In => {
                                let kind = if complete_side_expr
                                    .source_loc()
                                    .and_then(|loc| loc.snippet())
                                    .is_some_and(|snip| snip.contains('[') && snip.contains(']'))
                                {
                                    InKind::ArrayLiteral
                                } else {
                                    InKind::Entity
                                };
                                return CompletionContextKind::BinaryOp(BinaryOpContext {
                                    op: Op::In(kind),
                                    other_side_expr,
                                    complete_side_expr,
                                })
                                .into();
                            }
                            BinaryOp::Eq => {
                                return CompletionContextKind::BinaryOp(BinaryOpContext {
                                    op: Op::Eq,
                                    other_side_expr,
                                    complete_side_expr,
                                })
                                .into()
                            }
                            _ => (),
                        }
                    }
                }

                None
            })
    }

    fn visit_has_attr(
        &mut self,
        expr: &Arc<Expr>,
        attr: &SmolStr,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        self.visit_attr(expr, attr, loc, AttrContextKind::Has)
    }

    fn visit_get_attr(
        &mut self,
        expr: &Arc<Expr>,
        attr: &SmolStr,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        self.visit_attr(expr, attr, loc, AttrContextKind::Get)
    }

    fn visit_like(
        &mut self,
        expr: &Arc<Expr>,
        pattern: &Pattern,
        loc: Option<&Loc>,
    ) -> Option<Self::Output> {
        self.visit_expr(expr).or_else(|| {
            let word = self.cx.get_word_under_cursor();
            let pat = pattern.to_string();
            if self.cx.is_cursor_over_loc(loc)
                && (word == Some(pat.as_str()) || word.is_none() && pat.is_empty())
            {
                return CompletionContextKind::None.into();
            }
            None
        })
    }
}
