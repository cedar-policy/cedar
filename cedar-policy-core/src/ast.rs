/*
 * Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//! This module contains the AST datatypes.

mod expr;
pub use expr::*;
mod entity;
pub use entity::*;
mod extension;
pub use extension::*;
mod id;
pub use id::*;
mod integer;
pub use integer::{InputInteger, Integer};
mod literal;
pub use literal::*;
mod name;
pub use name::*;
mod ops;
pub use ops::*;
mod pattern;
pub use pattern::*;
mod partial_value;
pub use partial_value::*;
mod policy;
pub use policy::*;
mod policy_set;
pub use policy_set::*;
mod request;
pub use request::*;
mod restricted_expr;
pub use restricted_expr::*;
mod types;
pub use types::*;
mod value;
pub use value::*;
mod expr_iterator;
pub use expr_iterator::*;
