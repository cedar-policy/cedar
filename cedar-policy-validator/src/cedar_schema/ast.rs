/*
 * Copyright Cedar Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::iter::once;

use cedar_policy_core::{
    ast::{Id, InternalName},
    parser::{Loc, Node},
};
use itertools::{Either, Itertools};
use nonempty::NonEmpty;
use smol_str::SmolStr;
// We don't need this import on macOS but CI fails without it
#[allow(unused_imports)]
use smol_str::ToSmolStr;

use crate::json_schema;

pub const BUILTIN_TYPES: [&str; 3] = ["Long", "String", "Bool"];

pub(super) const CEDAR_NAMESPACE: &str = "__cedar";

pub type Schema = Vec<Node<Namespace>>;

/// A path is a non empty list of identifiers that forms a namespace + type
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Path(Node<PathInternal>);
impl Path {
    /// Create a [`Path`] with a single entry
    pub fn single(basename: Id, loc: Loc) -> Self {
        Self(Node::with_source_loc(
            PathInternal {
                basename,
                namespace: vec![],
            },
            loc,
        ))
    }

    /// Create [`Path`] with a head and an iterator. Most significant name first.
    pub fn new(basename: Id, namespace: impl IntoIterator<Item = Id>, loc: Loc) -> Self {
        let namespace = namespace.into_iter().collect();
        Self(Node::with_source_loc(
            PathInternal {
                basename,
                namespace,
            },
            loc,
        ))
    }

    /// Borrowed iteration of the [`Path`]'s elements. Most significant name first
    pub fn iter(&self) -> impl Iterator<Item = &Id> {
        self.0.node.iter()
    }

    /// Source [`Loc`] of this [`Path`]
    pub fn loc(&self) -> &Loc {
        &self.0.loc
    }

    /// Consume the [`Path`] and get an owned iterator over the elements. Most significant name first
    pub fn into_iter(self) -> impl Iterator<Item = Node<Id>> {
        let loc = self.0.loc;
        self.0
            .node
            .into_iter()
            .map(move |x| Node::with_source_loc(x, loc.clone()))
    }

    /// Get the base type name as well as the (potentially empty) prefix
    pub fn split_last(self) -> (Vec<Id>, Id) {
        (self.0.node.namespace, self.0.node.basename)
    }

    /// Is this referring to a name in the `__cedar` namespace (eg: `__cedar::Bool`)
    pub fn is_in_cedar(&self) -> bool {
        self.0.node.is_in_cedar()
    }
}

impl From<Path> for InternalName {
    fn from(value: Path) -> Self {
        InternalName::new(
            value.0.node.basename,
            value.0.node.namespace,
            Some(value.0.loc),
        )
    }
}

impl std::fmt::Display for Path {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0.node)
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct PathInternal {
    basename: Id,
    namespace: Vec<Id>,
}

impl PathInternal {
    fn iter(&self) -> impl Iterator<Item = &Id> {
        self.namespace.iter().chain(once(&self.basename))
    }

    fn into_iter(self) -> impl Iterator<Item = Id> {
        self.namespace.into_iter().chain(once(self.basename))
    }

    /// Is this referring to a name _in_ the `__cedar` namespace (ex: `__cedar::Bool`)
    fn is_in_cedar(&self) -> bool {
        match self.namespace.as_slice() {
            [id] => id.as_ref() == CEDAR_NAMESPACE,
            _ => false,
        }
    }
}

impl std::fmt::Display for PathInternal {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if self.namespace.is_empty() {
            write!(f, "{}", self.basename)
        } else {
            let namespace = self.namespace.iter().map(|id| id.as_ref()).join("::");
            write!(f, "{namespace}::{}", self.basename)
        }
    }
}

/// This struct represents Entity Uids in the Schema Syntax
#[derive(Debug, Clone)]
pub struct QualName {
    pub path: Option<Path>,
    pub eid: SmolStr,
}

impl QualName {
    pub fn unqualified(eid: SmolStr) -> Self {
        Self { path: None, eid }
    }

    pub fn qualified(path: Path, eid: SmolStr) -> Self {
        Self {
            path: Some(path),
            eid,
        }
    }
}

/// A [`Namespace`] has a name and a collection declaration
/// A schema is made up of a series of fragments
/// A fragment is a series of namespaces
#[derive(Debug, Clone)]
pub struct Namespace {
    /// The name of this namespace. If [`None`], then this is the unqualified namespace
    pub name: Option<Node<Path>>,
    /// The [`Declaration`]s contained in this namespace
    pub decls: Vec<Node<Declaration>>,
}

impl Namespace {
    /// Is this [`Namespace`] unqualfied?
    pub fn is_unqualified(&self) -> bool {
        self.name.is_none()
    }
}

pub trait Decl {
    fn names(&self) -> Vec<Node<SmolStr>>;
}

/// Schema Declarations,
/// Defines either entity types, action types, or common types
#[derive(Debug, Clone)]
pub enum Declaration {
    Entity(EntityDecl),
    Action(ActionDecl),
    Type(TypeDecl),
}

#[derive(Debug, Clone)]
pub struct TypeDecl {
    pub name: Node<Id>,
    pub def: Node<Type>,
}

impl Decl for TypeDecl {
    fn names(&self) -> Vec<Node<SmolStr>> {
        vec![self.name.clone().map(|id| id.to_smolstr())]
    }
}

/// Declaration of an entity type
#[derive(Debug, Clone)]
pub struct EntityDecl {
    /// Entity Type Names bound by this declaration.
    /// More than one name can be bound if they have the same definition, for convenience
    pub names: Vec<Node<Id>>,
    /// Entity Types this type is allowed to be related to via the `in` relation
    pub member_of_types: Vec<Path>,
    /// Attributes this entity has
    pub attrs: Vec<Node<AttrDecl>>,
}

/// Type definitions
#[derive(Debug, Clone)]
pub enum Type {
    /// A set of types
    Set(Box<Node<Type>>),
    /// A [`Path`] that could either refer to a Common Type or an Entity Type
    Ident(Path),
    /// A Record
    Record(Vec<Node<AttrDecl>>),
}

/// Primitive Type Definitions
#[derive(Debug, Clone)]
pub enum PrimitiveType {
    /// Cedar Longs
    Long,
    /// Cedar Strings
    String,
    /// Cedar booleans
    Bool,
}

impl<N> From<PrimitiveType> for json_schema::TypeVariant<N> {
    fn from(value: PrimitiveType) -> Self {
        match value {
            PrimitiveType::Long => json_schema::TypeVariant::Long,
            PrimitiveType::String => json_schema::TypeVariant::String,
            PrimitiveType::Bool => json_schema::TypeVariant::Boolean,
        }
    }
}

/// Attribute declarations , used in records and entity types
#[derive(Debug, Clone)]
pub struct AttrDecl {
    /// Name of this attribute
    pub name: Node<SmolStr>,
    /// Whether or not it is a required attribute (implicitly `true`)
    pub required: bool,
    /// The type of this attribute
    pub ty: Node<Type>,
}

/// The target of a [`PRAppDecl`]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PR {
    /// Applies to the `principal` variable
    Principal,
    /// Applies to the `resource` variable
    Resource,
}

impl std::fmt::Display for PR {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PR::Principal => write!(f, "principal"),
            PR::Resource => write!(f, "resource"),
        }
    }
}

/// A declaration that defines what kind of entities this action can be applied against
#[derive(Debug, Clone)]
pub struct PRAppDecl {
    /// Is this constraining the `principal` or the `resource`
    pub kind: Node<PR>,
    /// What entity types are allowed?
    pub entity_tys: NonEmpty<Path>,
}

/// A declaration of constraints on an action type
#[derive(Debug, Clone)]
pub enum AppDecl {
    /// Constraints on the `principal` or `resource`
    PR(PRAppDecl),
    /// Constraints on the `context`
    Context(Either<Path, Vec<Node<AttrDecl>>>),
}

/// An action declaration
#[derive(Debug, Clone)]
pub struct ActionDecl {
    /// The names this declaration is binding.
    /// More than one name can be bound if they have the same definition, for convenience.
    pub names: NonEmpty<Node<SmolStr>>,
    /// The parents of this action
    pub parents: Option<NonEmpty<Node<QualName>>>,
    /// The constraining clauses in this declarations
    pub app_decls: Option<Node<NonEmpty<Node<AppDecl>>>>,
}

impl Decl for ActionDecl {
    fn names(&self) -> Vec<Node<SmolStr>> {
        self.names.iter().cloned().collect()
    }
}

#[cfg(test)]
mod test {
    use std::sync::Arc;

    use super::*;

    fn loc() -> Loc {
        Loc::new((1, 1), Arc::from("foo"))
    }

    // Ensure the iterators over [`Path`]s return most significant names first
    #[test]
    fn path_iter() {
        let p = Path::new(
            "baz".parse().unwrap(),
            ["foo".parse().unwrap(), "bar".parse().unwrap()],
            loc(),
        );

        let expected: Vec<Id> = vec![
            "foo".parse().unwrap(),
            "bar".parse().unwrap(),
            "baz".parse().unwrap(),
        ];

        let expected_borrowed = expected.iter().collect::<Vec<_>>();

        let borrowed = p.iter().collect::<Vec<_>>();
        assert_eq!(borrowed, expected_borrowed);
        let moved = p.into_iter().map(|n| n.node).collect::<Vec<_>>();
        assert_eq!(moved, expected);
    }
}
