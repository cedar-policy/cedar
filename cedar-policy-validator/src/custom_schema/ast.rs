use cedar_policy_core::{
    ast::{Id, Name as CName},
    parser::Node,
};
use itertools::Either;
use nonempty::NonEmpty;
use smol_str::SmolStr;

const IPADDR_EXTENSION: &str = "ipaddr";
const DECIMAL_EXTENSION: &str = "decimal";
const CEDAR_NAMESPACE: &str = "__cedar";

pub type Ident = Node<Id>;
pub type Str = Node<SmolStr>;
pub type Name = Either<Ident, Str>;
pub type Schema = Vec<Node<Namespace>>;

#[derive(Debug, Clone)]
pub struct Path {
    pub base: Ident,
    pub prefix: Vec<Ident>,
}

impl Path {
    pub fn is_ipaddr_extension(&self) -> bool {
        self.is_cedar_builtin() && self.base.node.clone().to_smolstr() == IPADDR_EXTENSION
    }

    pub fn is_decimal_extension(&self) -> bool {
        self.is_cedar_builtin() && self.base.node.clone().to_smolstr() == DECIMAL_EXTENSION
    }

    pub fn to_smolstr(self) -> SmolStr {
        <Path as Into<CName>>::into(self).to_string().into()
    }

    pub fn is_unqualified_ipaddr(&self) -> bool {
        self.prefix.is_empty() && self.base.node.clone().to_smolstr() == IPADDR_EXTENSION
    }

    pub fn is_unqualified_decimal(&self) -> bool {
        self.prefix.is_empty() && self.base.node.clone().to_smolstr() == DECIMAL_EXTENSION
    }

    fn is_cedar_builtin(&self) -> bool {
        matches!(self.prefix.as_slice(), [Node { node, loc: _ }] if node.as_ref() == CEDAR_NAMESPACE)
    }
}

// Convert `Path` to `cedar_policy_core::ast::Name` and drop all its source locations
impl From<Path> for CName {
    fn from(value: Path) -> Self {
        CName::new(
            value.base.node,
            value.prefix.iter().map(|id| id.node.clone()),
        )
    }
}

#[derive(Debug, Clone)]
pub struct Namespace {
    pub name: Option<Node<Path>>,
    pub decls: Vec<Node<Declaration>>,
}

#[derive(Debug, Clone)]
pub enum Declaration {
    Entity(EntityDecl),
    Action(ActionDecl),
    Type(Ident, Node<Type>),
}

#[derive(Debug, Clone)]
pub struct EntityDecl {
    pub names: Vec<Ident>,
    pub member_of_types: Option<Vec<Node<Path>>>,
    pub attrs: Vec<Node<AttrDecl>>,
}

#[derive(Debug, Clone)]
pub enum Type {
    Long,
    Bool,
    String,
    Set(Box<Node<Type>>),
    Ident(Path),
    Record(Vec<Node<AttrDecl>>),
}

#[derive(Debug, Clone)]
pub struct AttrDecl {
    pub name: Name,
    pub required: Option<()>,
    pub ty: Node<Type>,
}

#[derive(Debug, Clone)]
pub enum PR {
    Principal,
    Resource,
}

#[derive(Debug, Clone)]
pub struct PRAppDecl {
    pub ty: Node<PR>,
    pub entity_tys: NonEmpty<Node<Path>>,
}

#[derive(Debug, Clone)]
pub enum AppDecl {
    PR(PRAppDecl),
    Context(Vec<Node<AttrDecl>>),
}

#[derive(Debug, Clone)]
pub struct Ref {
    pub ty: Option<Path>,
    pub id: Name,
}

#[derive(Debug, Clone)]
pub struct ActionDecl {
    pub names: Vec<Name>,
    pub parents: Option<Vec<Ref>>,
    pub app_decls: Option<NonEmpty<Node<AppDecl>>>,
}
