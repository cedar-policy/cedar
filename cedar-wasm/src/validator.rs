use cedar_policy::frontend::{utils::InterfaceResult, validate::json_validate};
use wasm_bindgen::prelude::*;

#[wasm_bindgen(js_name = "validate")]
pub fn wasm_validate(input: &str) -> InterfaceResult {
    json_validate(input)
}
