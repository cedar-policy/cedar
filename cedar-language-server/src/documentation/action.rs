/*
 * Copyright Cedar Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use std::borrow::Cow;

use cedar_policy_core::ast::{ActionConstraint, EntityUID};

use super::ToDocumentationString;
use crate::{
    markdown::MarkdownBuilder,
    policy::{format_attributes, DocumentContext},
};
use cedar_policy_core::validator::{
    types::{EntityRecordKind, Type},
    ValidatorSchema,
};
use itertools::Itertools;

pub(crate) struct ActionDocumentation<'a> {
    constraint: Option<&'a ActionConstraint>,
}

impl<'a> ActionDocumentation<'a> {
    pub(crate) fn new(constraint: Option<&'a ActionConstraint>) -> Self {
        Self { constraint }
    }
}

impl<'a> From<&'a ActionConstraint> for ActionDocumentation<'a> {
    fn from(value: &'a ActionConstraint) -> Self {
        Self::new(Some(value))
    }
}

impl<'a> From<&'a DocumentContext<'_>> for ActionDocumentation<'a> {
    fn from(value: &'a DocumentContext<'_>) -> Self {
        value.policy.action_constraint().into()
    }
}

impl ToDocumentationString for ActionDocumentation<'_> {
    fn to_documentation_string(&self, schema: Option<&ValidatorSchema>) -> Cow<'static, str> {
        let static_docs = include_str!("markdown/action.md");
        let Some(constraint) = &self.constraint else {
            return static_docs.into();
        };
        let mut builder = MarkdownBuilder::new();
        builder.push_str(static_docs);

        match constraint {
            ActionConstraint::Any => {
                builder.header("Available Actions");
                if let Some(schema) = schema {
                    format_action_list(&mut builder, schema.actions().sorted());
                } else {
                    builder.paragraph("*Schema not available - any action permitted*");
                }
            }
            ActionConstraint::In(entity_uids) => {
                builder
                    .header("Permitted Actions")
                    .paragraph("This policy applies to the following actions:");

                let entity_uids = entity_uids.iter().sorted().map(AsRef::as_ref);
                format_action_list(&mut builder, entity_uids);
            }
            ActionConstraint::Eq(entity_uid) => {
                builder
                    .header("Specific Action")
                    .paragraph("This policy applies only to the following action:")
                    .code_block("cedar", &entity_uid.to_string());

                // Add context details from schema if available
                if let Some(schema) = schema {
                    if let Some(action) = schema.get_action_id(entity_uid) {
                        if let Type::EntityOrRecord(EntityRecordKind::Record { attrs, .. }) =
                            action.context()
                        {
                            builder
                                .header("Context Attributes")
                                .code_block("cedarschema", &format_attributes(attrs));
                        }
                    }
                }
            }
            ActionConstraint::ErrorConstraint => {
                builder
                    .header("Error")
                    .paragraph("Invalid action constraint");
            }
        }

        builder.build().into()
    }
}

fn format_action_list<'a>(
    builder: &mut MarkdownBuilder,
    actions: impl ExactSizeIterator<Item = &'a EntityUID>,
) {
    const MAX_ACTIONS_TO_SHOW: usize = 10;
    let actions_len = actions.len();
    if actions_len == 0 {
        builder.paragraph("*No actions specified*");
        return;
    }

    let shown_actions = actions.take(MAX_ACTIONS_TO_SHOW);
    let remaining = if shown_actions.len() == MAX_ACTIONS_TO_SHOW {
        actions_len - MAX_ACTIONS_TO_SHOW
    } else {
        0
    };

    builder.code_block("cedar", &shown_actions.map(ToString::to_string).join("\n"));

    if remaining > 0 {
        builder.paragraph(&format!("*... and {remaining} more actions defined*"));
    }
}
