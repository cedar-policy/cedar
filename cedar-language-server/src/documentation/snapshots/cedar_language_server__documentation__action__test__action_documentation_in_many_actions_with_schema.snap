---
source: cedar-language-server/src/documentation/action.rs
expression: action_doc.to_documentation_string(Some(&schema))
---
# Action

The action element in a Cedar policy is the action that can be performed on the resource
defined by the resource element.

The action element must be present. If you specify only action without an expression
that constrains its scope, then the policy applies to any action.

**Permitted Actions**

This policy applies to the following actions:

```cedar
Action::"Create"
Action::"Delete"
Action::"Download"
Action::"Edit"
Action::"List"
Action::"Read"
Action::"Search"
Action::"Share"
Action::"Update"
Action::"Upload"
```

*... and 2 more actions defined*
